### PR TITLE
Feat: Toast notification timeout settings

### LINKED ISSUE #841: [Feature]: Clear All Toasts
### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

Within the Web App, if there were many detected down times, the UI fills up with red Toasts that must be dismissed one by one.  Green 'up' messages timeout on their own, but the red 'down' messages remain.  It would be a good UX improvement to provide a 'close all' button.

### ✔️ Solution

A small semi-transparent X at the bottom right corner to clear all Toast messages.

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar feature request

### LINKED ISSUE #1085: Autoclose alert
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

Autoclose alert popup after a defined time. This is useful if the page is displayed on a TV screen for example. 

### ✔️ Solution

Add a timeout option in the configuration. Default timeout to 1min ?

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### LINKED ISSUE #1178: Auto Dismissing Down (Red) Notifications
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

Have an option to change Down Notifications to the Auto Dismiss type like the Up Notifications do automatically.

### ✔️ Solution

When a "Down" (Red) Notification appears on screen it auto dismisses when items comes back online.

### ❓ Alternatives

Maybe a toggle in settings to choose between "Auto Dismiss" or "Persistant" 

### 📝 Additional Context

_No response_

### LINKED ISSUE #2116: Popup: timestamp, timeout
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

You should add:
 - timestamp/(x minutes ago?) in the popup (When did we have the event?) This would be important
 - option for auto-disappearing the error-popups. (Generic+monitor specific option as well)
 - Button for deleting all popups (it's very annoying to close the popups one by one)

### ✔️ Solution

The feature description contains all the needed info about the solution, I guess. ;)

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by chakflying**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description

Fixes #2116
Fixes #1178
Fixes #1085
Fixes #841

This is a continuation of #2555, with various changes applied to improve usability and code organization.

Settings is stored to `localStorage` on change, so the "Save" button is not needed. Settings are applied to the toast instance default, which will apply to all toast notifications of the same type.

A new button to clear all toast notification is added to the bottom of the toasts stack. The individual close button for each toast is set to show on hover since the close all button should satisfy most needs.

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

image

image

### CODE REVIEW

**Comment by chakflying** [comment]
Thinking about this, it's probably not a good idea for the setting to affect the system messages. Changed back to the setting affecting monitor toast notifications only.

**Comment by CommanderStorm** [comment]
Could you add the following to the PR description?
Resolves https://github.com/louislam/uptime-kuma/issues/1085
Resolves https://github.com/louislam/uptime-kuma/issues/841

File: src/util-frontend.js [hunk 1]
```diff
@@ -139,3 +140,58 @@ export function colorOptions(self) {
...
+            if (toast.timeout === 0) {
+                return false;
+            } else {
+                return toast;
+            }
+        },
+    };
+}
+
+export function getToastSuccessTimeout() {
```
Comment (Computroniks) [inline src/util-frontend.js:165]:
```suggestion
/**
 * Get timeout for success toasts
 * @returns {(number|boolean)} Timeout in ms. If false timeout disabled.
 */
export function getToastSuccessTimeout() {
```

File: src/util-frontend.js [hunk 2]
```diff
@@ -139,3 +140,58 @@ export function colorOptions(self) {
...
+    }
+
+    if (successTimeout === -1) {
+        successTimeout = false;
+    }
+
+    return successTimeout;
+}
+
+export function getToastErrorTimeout() {
```
Comment (Computroniks) [inline src/util-frontend.js:182]:
```suggestion
/**
 * Get timeout for error toasts
 * @returns {(number|boolean)} Timeout in ms. If false timeout disabled.
 */
export function getToastErrorTimeout() {
```

File: src/components/settings/Notifications.vue [hunk 3]
```diff
@@ -77,6 +112,26 @@ export default {
         },
     },
 
+    watch: {
+        // Parse, store and apply new timeout settings.
+        toastSuccessTimeoutSecs: function (newTimeout) {
```
Comment (Computroniks) [inline src/components/settings/Notifications.vue:117]:
```suggestion
        toastSuccessTimeoutSecs(newTimeout) {
```

This seems to fit better with how it is styled in the rest of the code base and in the VueJS docs.

File: src/components/settings/Notifications.vue [hunk 4]
```diff
@@ -77,6 +112,26 @@ export default {
...
+            const parsedTimeout = parseInt(newTimeout);
+            if (parsedTimeout != null && !Number.isNaN(parsedTimeout)) {
+                localStorage.toastSuccessTimeout = newTimeout > 0 ? newTimeout * 1000 : newTimeout;
+            }
+        },
+        toastErrorTimeoutSecs: function (newTimeout) {
```
Comment (Computroniks) [inline src/components/settings/Notifications.vue:123]:
```suggestion
        toastErrorTimeoutSecs(newTimeout) {
```

File: src/util-frontend.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import dayjs from "dayjs";
 import timezones from "timezones-list";
 import { localeDirection, currentLocale } from "./i18n";
+import { TYPE, POSITION } from "vue-toastification";
```
Comment (Computroniks) [inline src/util-frontend.js:4]:
```suggestion
import { POSITION } from "vue-toastification";
```

Linter warning

**Review by Computroniks — CHANGES_REQUESTED** [review]

**Comment by domdfcoding** [comment]
Are there any instructions for building the docker image from source? I looked but I can only find npm instructions, and my current setup is through docker. I'd like to give this PR a try as it's a feature I'm looking forward to.

**Comment by Computroniks** [comment]
You should be able to use the pre built docker image for this PR by following these instructions in the wiki
https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

**Comment by mezza93** [comment]
> Are there any instructions for building the docker image from source? I looked but I can only find npm instructions, and my current setup is through docker. I'd like to give this PR a try as it's a feature I'm looking forward to.

Sorry but where did you find the instructions for NPM upgrade for this one?

**Comment by CommanderStorm** [comment]
See #4500 for the steps where help is appreciated to get v2.0 shipped

**Pull request merged**